plugin/advisor/catalog/walk_through.go

```diff
@@ -0,0 +1,420 @@
+	ErrorTypeColumnExists = 401
+	// ErrorTypeColumnNotExists is the error that column not exists.
+	ErrorTypeColumnNotExists = 504
+
+	// 501 ~ 599 index error type.
+
+	// ErrorTypePrimaryKeyExists is the error that PK exists.
+	ErrorTypePrimaryKeyExists = 501
+	// ErrorTypeIndexExists is the error that index exists.
+	ErrorTypeIndexExists = 502
+	// ErrorTypeIndexEmptyKeys is the error that index has empty keys.
+	ErrorTypeIndexEmptyKeys = 503
+)
+
+// WalkThroughError is the error for walking-through.
+type WalkThroughError struct {
+	Type    WalkThroughErrorType
+	Content string
+}
+
+// NewParseError returns a new ErrorTypeParseError.
+func NewParseError(content string) *WalkThroughError {
+	return &WalkThroughError{
+		Type:    ErrorTypeParseError,
+		Content: content,
+	}
+}
+
+// Error implements the error interface.
+func (e *WalkThroughError) Error() string {
+	return e.Content
+}
+
+// WalkThrough will change the catalog schema by statements.
+func (d *databaseState) WalkThrough(stmts string) error {
+	if d.dbType != db.MySQL {
+		return nil
+	}
+
+	// For MySQL, there is only one schema whose name is empty string.
+	// If there is no empty-string-name schema, create it to avoid corner cases.
+	if _, exists := d.schemaSet[""]; !exists {
+		d.createSchema("")
+	}
+
+	nodeList, err := d.parse(stmts)
+	if err != nil {
+		return err
+	}
+
+	for _, node := range nodeList {
+		if err := d.changeState(node); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (d *databaseState) changeState(in tidbast.StmtNode) error {
+	switch node := in.(type) {
+	case *tidbast.CreateTableStmt:
+		return d.createTable(node)
+	default:
+		return nil
+	}
+}
+
+func (d *databaseState) createTable(node *tidbast.CreateTableStmt) error {
+	if node.Table.Schema.O != "" && d.name != node.Table.Schema.O {
+		return &WalkThroughError{
+			Type:    ErrorTypeAccessOtherDatabase,
+			Content: fmt.Sprintf("Database `%s` is not the current database", node.Table.Schema.O),
+		}
+	}
+
+	schema, exists := d.schemaSet[""]
+	if !exists {
+		schema = d.createSchema("")
+	}
+
+	if _, exists = schema.tableSet[node.Table.Name.O]; exists {
+		if node.IfNotExists {
+			return nil
+		}
+		return &WalkThroughError{
+			Type:    ErrorTypeTableExists,
+			Content: fmt.Sprintf("Table `%s` exists", node.Table.Name.O),
+		}
+	}
+
+	table := &tableState{
+		name:      node.Table.Name.O,
+		columnSet: make(columnStateMap),
+		indexSet:  make(indexStateMap),
+	}
+	schema.tableSet[table.name] = table
+
+	for i, column := range node.Cols {
+		if err := table.createColumn(column, i+1); err != nil {
+			return err
+		}
+	}
+
+	for _, constraint := range node.Constraints {
+		if err := table.createConstraint(constraint); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (t *tableState) createConstraint(constraint *tidbast.Constraint) error {
+	switch constraint.Tp {
+	case tidbast.ConstraintPrimaryKey:
+		keyList, err := t.validateAndGetKeyStringList(constraint.Keys, true)
```

Comment:
For readability, boolean normally would put an inline comment

t.validateAndGetKeyStringList(constraint.Keys, true /\*primary\*/)